### ISSUE TITLE
[basic.def.odr] Example seemingly constradicts normative wording

### ISSUE THREAD

**Issue by Eisenwave** [cplusplus/draft#6364]
On one hand, lambda-expressions in default arguments are not required to have the same closure type:

> In each such definition, except within the default arguments and default template arguments of D, corresponding [lambda-expression](http://eel.is/c++draft/expr.prim.lambda.general#nt:lambda-expression)s shall have the same closure type (see below).

http://eel.is/c++draft/basic.def.odr#14.6

On the other hand, the example shows the opposite:
```cpp
inline void g(bool cond, void (*p)() = []{}) {
  if (cond) g(false);
}
```
> If the definition of g appears in multiple translation units, the program is ill-formed (no diagnostic required) because each such definition uses a default argument that refers to a distinct [lambda-expression](http://eel.is/c++draft/expr.prim.lambda.general#nt:lambda-expression) closure type.

http://eel.is/c++draft/basic.def.odr#16

I can't make sense of it; it seems like a direct contradiction to me.

**Comment by frederick-vs-ja** [comment]
I think p14.6 specifies behavior of implementations, while the example shows potentially invalid use in user codes due to such behavior, so there's no contradiction.

**Comment by Eisenwave** [comment]
> I think p14.6 specifies behavior of implementations, while the example shows potentially invalid use in user codes due to such behavior, so there's no contradiction.

My issue is that 14.6 says **except within the default arguments**, so it seemingly excludes the behavior described in the example.

**Comment by JohelEGP** [comment]
> **except within the default arguments**

It's saying that the closure types are the same _except_ in that case.

**Comment by Eisenwave** [comment]
> > **except within the default arguments**
> 
> It's saying that the closure types are the same _except_ in that case.

It's not saying that; the quote I've posted is just one of many requirements for the ODR not to be violated.

**Comment by Eisenwave** [comment]
After discussing it with a handful of other people, we managed to wrap our heads around it. A much better example would be:
```cpp
inline void g(void (*p)() = []{}) {}

inline void g2() {
    // ill-formed, no diagnostic required
    // use of default argument defines a closure type outside of g
    // which is not the same across TUs
    g();
}
```

**Comment by JohelEGP** [comment]
So the current example also requires <http://eel.is/c++draft/basic.def.odr#14.11> to make sense.
Is that right?

**Comment by Eisenwave** [comment]
> So the current example also requires http://eel.is/c++draft/basic.def.odr#14.11 to make sense. Is that right?

Not really. `g` making that recursive call means that `[]{}` is introduced into its definition for ODR purposes, but that doesn't create any additional problems. We violate the ODR with lambdas when they don't have the same closure type everywhere; that's the actual issue.

It may look like p14.11 is relevant to the example with `g`, but it isn't at all.
